File: backend/app/routes/textures.py
from __future__ import annotations
from fastapi import APIRouter, HTTPException, UploadFile, File, Form
from fastapi.responses import FileResponse
from pydantic import BaseModel
from typing import Optional

from app.models.textures import TextureAsset, MaterialTextureSet, VALID_TEXTURE_TYPES
from app.services import texture_service
from app.services.project_context import get_textures_dir, get_active_project_id

router = APIRouter(prefix="/api/textures", tags=["textures"])


class AssignBody(BaseModel):
    material_profile_id: str
    textures: dict[str, str]   # texture_type → texture_id
    asset_id: Optional[str] = None


@router.get("", response_model=list[TextureAsset])
def list_textures():
    return texture_service.list_textures()


@router.post("/upload", response_model=TextureAsset, status_code=201)
async def upload_texture(
    file: UploadFile = File(...),
    texture_type: str = Form("albedo"),
    name: str = Form(""),
    tags: str = Form(""),
):
    if texture_type not in VALID_TEXTURE_TYPES:
        raise HTTPException(status_code=422, detail=f"Invalid texture_type. Allowed: {sorted(VALID_TEXTURE_TYPES)}")

    content = await file.read()
    tag_list = [t.strip() for t in tags.split(",") if t.strip()] if tags else []

    try:
        asset = texture_service.save_texture_file(
            file_bytes=content,
            original_filename=file.filename or "texture.png",
            texture_type=texture_type,
            name=name,
            tags=tag_list,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))

    return asset


@router.delete("/{texture_id}")
def delete_texture(texture_id: str):
    removed = texture_service.delete_texture(texture_id)
    if not removed:
        raise HTTPException(status_code=404, detail="Texture not found")
    return {"deleted": texture_id}


@router.post("/assign", response_model=MaterialTextureSet)
def assign_textures(body: AssignBody):
    try:
        result = texture_service.assign_textures(
            material_profile_id=body.material_profile_id,
            textures=body.textures,
            asset_id=body.asset_id,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))
    return result


@router.get("/assignment/{material_profile_id}", response_model=MaterialTextureSet)
def get_assignment(material_profile_id: str):
    result = texture_service.get_assignment(material_profile_id)
    if result is None:
        raise HTTPException(status_code=404, detail="No assignment found for this material profile")
    return result


@router.get("/file/{texture_id}")
def serve_texture_file(texture_id: str):
    """Serve raw texture file by texture ID (active project)."""
    path = texture_service.get_texture_file_path(texture_id)
    if path is None or not path.exists():
        raise HTTPException(status_code=404, detail="Texture file not found")
    return FileResponse(str(path))

File: frontend/src/lib/textures.ts
export const TEXTURE_TYPES = ["albedo", "normal", "roughness", "metallic", "emissive", "opacity", "ao"] as const;
export type TextureType = (typeof TEXTURE_TYPES)[number];

export const TEXTURE_TYPE_LABELS: Record<TextureType, string> = {
  albedo: "Albedo",
  normal: "Normal",
  roughness: "Roughness",
  metallic: "Metallic",
  emissive: "Emissive",
  opacity: "Opacity",
  ao: "AO",
};

export interface TextureAsset {
  id: string;
  project_id: string;
  name: string;
  texture_type: TextureType;
  filename: string;
  file_size: number;
  resolution: string | null;
  created_at: string;
  tags: string[];
}

export interface MaterialTextureSet {
  id: string;
  asset_id: string | null;
  material_profile_id: string;
  assigned_textures: Partial<Record<TextureType, string>>;  // texture_type → texture_id
  created_at: string;
  updated_at: string;
}

export function textureUrl(asset: TextureAsset): string {
  return `/textures/${asset.filename}`;
}

export async function listTextures(): Promise<TextureAsset[]> {
  const res = await fetch("/api/textures");
  if (!res.ok) throw new Error("Failed to fetch textures");
  return res.json();
}

export async function uploadTexture(
  file: File,
  opts: { texture_type: TextureType; name?: string; tags?: string[] }
): Promise<TextureAsset> {
  const fd = new FormData();
  fd.append("file", file);
  fd.append("texture_type", opts.texture_type);
  if (opts.name) fd.append("name", opts.name);
  if (opts.tags?.length) fd.append("tags", opts.tags.join(","));

  const res = await fetch("/api/textures/upload", { method: "POST", body: fd });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Upload failed" }));
    throw new Error(err.detail ?? "Upload failed");
  }
  return res.json();
}

export async function deleteTexture(textureId: string): Promise<void> {
  const res = await fetch(`/api/textures/${textureId}`, { method: "DELETE" });
  if (!res.ok) throw new Error("Failed to delete texture");
}

export async function assignTextures(
  materialProfileId: string,
  textures: Partial<Record<TextureType, string>>,
  assetId?: string
): Promise<MaterialTextureSet> {
  const res = await fetch("/api/textures/assign", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      material_profile_id: materialProfileId,
      textures,
      asset_id: assetId ?? null,
    }),
  });
  if (!res.ok) throw new Error("Failed to assign textures");
  return res.json();
}

export async function getAssignment(materialProfileId: string): Promise<MaterialTextureSet | null> {
  const res = await fetch(`/api/textures/assignment/${materialProfileId}`);
  if (res.status === 404) return null;
  if (!res.ok) throw new Error("Failed to get assignment");
  return res.json();
}

export function formatFileSize(bytes: number): string {
  if (bytes === 0) return "0 B";
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}
